### PR TITLE
Split NLSolve macros

### PR DESCRIPTION
**Pull request by kanav99**
In this we I have split the previously macros `oopnlsolve` and `iipnlsolve` into three parts -
1. Functions `iipnlsolve` and `oopnlsolve` which make the `nlsolver` using the cache arguments.
2. Functions `iip_generate_W` and `oop_generate_W` which make the `W` operator using cache arguments. I had to make a seperate function because `WOperator` is not defined inside DiffEqBase.
3. Macros `@getiipnlsolvefields` and `@getoopnlsolvefields` to get other fields used in algorithm caches of algos which used the previous macros.
Right now I get some instability issues, trying to resolve them.

### CODE REVIEW

**Comment by kanav99** [comment]
I wonder why SSPRK error, I get this error here https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/blob/master/test/algconvergence/ode_ssprk_tests.jl#L58

```
ERROR: LoadError: MethodError: no method matching iterate(::Nothing)
Closest candidates are:
  iterate(!Matched::Core.SimpleVector) at essentials.jl:568
  iterate(!Matched::Core.SimpleVector, !Matched::Any) at essentials.jl:568
  iterate(!Matched::ExponentialBackOff) at error.jl:199
  ...
Stacktrace:
 [1] isempty(::Nothing) at ./essentials.jl:766
 [2] #ConvergenceSimulation#57(::Dict{Symbol,Array{Rational{Int64},1}}, ::Nothing, ::Type, ::Array{ODESolution{Float64,1,Array{Float64,1},Nothing,Nothing,Array{Float64,1},Array{Array{Float64,1},1},ODEProblem{Float64,Tuple{Float64,Float64},false,Nothing,ODEFunction{false,getfield(Main, Symbol("##3#4")),LinearAlgebra.UniformScaling{Bool},Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing},Nothing,DiffEqBase.StandardODEProblem},SSPRK22{typeof(OrdinaryDiffEq.trivial_limiter!),typeof(OrdinaryDiffEq.trivial_limiter!)},OrdinaryDiffEq.InterpolationData{ODEFunction{false,getfield(Main, Symbol("##3#4")),LinearAlgebra.UniformScaling{Bool},Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing},Array{Float64,1},Array{Float64,1},Array{Array{Float64,1},1},OrdinaryDiffEq.SSPRK22ConstantCache},DiffEqBase.DEStats},1}, ::Array{Rational{Int64},1}) at /home/kanav/.julia/packages/DiffEqDevTools/l9UHq/src/convergence.jl:15
 [3] (::getfield(Core, Symbol("#kw#Type")))(::NamedTuple{(:auxdata,),Tuple{Dict{Symbol,Array{Rational{Int64},1}}}}, ::Type{ConvergenceSimulation}, ::Array{ODESolution{Float64,1,Array{Float64,1},Nothing,Nothing,Array{Float64,1},Array{Array{Float64,1},1},ODEProblem{Float64,Tuple{Float64,Float64},false,Nothing,ODEFunction{false,getfield(Main, Symbol("##3#4")),LinearAlgebra.UniformScaling{Bool},Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing},Nothing,DiffEqBase.StandardODEProblem},SSPRK22{typeof(OrdinaryDiffEq.trivial_limiter!),typeof(OrdinaryDiffEq.trivial_limiter!)},OrdinaryDiffEq.InterpolationData{ODEFunction{false,getfield(Main, Symbol("##3#4")),LinearAlgebra.UniformScaling{Bool},Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing},Array{Float64,1},Array{Float64,1},Array{Array{Float64,1},1},OrdinaryDiffEq.SSPRK22ConstantCache},DiffEqBase.DEStats},1}, ::Array{Rational{Int64},1}) at ./none:0
 [4] #test_convergence#78(::Bool, ::Bool, ::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::Function, ::Array{Rational{Int64},1}, ::ODEProblem{Float64,Tuple{Float64,Float64},false,Nothing,ODEFunction{false,getfield(Main, Symbol("##3#4")),LinearAlgebra.UniformScaling{Bool},Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing},Nothing,DiffEqBase.StandardODEProblem}, ::SSPRK22{typeof(OrdinaryDiffEq.trivial_limiter!),typeof(OrdinaryDiffEq.trivial_limiter!)}) at /home/kanav/.julia/packages/DiffEqDevTools/l9UHq/src/convergence.jl:98
 [5] test_convergence(::Array{Rational{Int64},1}, ::ODEProblem{Float64,Tuple{Float64,Float64},false,Nothing,ODEFunction{false,getfield(Main, Symbol("##3#4")),LinearAlgebra.UniformScaling{Bool},Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing},Nothing,DiffEqBase.StandardODEProblem}, ::SSPRK22{typeof(OrdinaryDiffEq.trivial_limiter!),typeof(OrdinaryDiffEq.trivial_limiter!)}) at /home/kanav/.julia/packages/DiffEqDevTools/l9UHq/src/convergence.jl:95
 [6] top-level scope at /home/kanav/.julia/dev/OrdinaryDiffEq/test/algconvergence/ode_ssprk_tests.jl:58 [inlined]
 [7] top-level scope at ./none:0
 [8] include at ./boot.jl:326 [inlined]
 [9] include_relative(::Module, ::String) at ./loading.jl:1038
 [10] include(::Module, ::String) at ./sysimg.jl:29
 [11] exec_options(::Base.JLOptions) at ./client.jl:267
 [12] _start() at ./client.jl:436
in expression starting at /home/kanav/.julia/dev/OrdinaryDiffEq/test/algconvergence/ode_ssprk_tests.jl:57
```

**Comment by kanav99** [comment]
Also I get similar error at random times https://travis-ci.org/JuliaDiffEq/OrdinaryDiffEq.jl/jobs/549693439

**Comment by kanav99** [comment]
Updates: The errors were fixed in the OrdinaryDiffEq PR

File: src/nlsolve/utils.jl [hunk 1]
```diff
@@ -102,25 +105,7 @@ DiffEqBase.@def iipnlsolve begin
...
-      if DiffEqBase.has_jac(f) && !DiffEqBase.has_invW(f) && f.jac_prototype !== nothing
-        W = WOperator(f, dt, true)
-        J = nothing # is J = W.J better?
-      else
-        J = false .* vec(u) .* vec(u)'
-        W = similar(J)
-      end
-    end
-
+    W = iip_generate_W(alg,u,uprev,p,t,dt,f,uEltypeNoUnits)
```
Comment (ChrisRackauckas) [inline src/nlsolve/utils.jl]: return J?

Comment (kanav99) [reply src/nlsolve/utils.jl]: J is not required there, J is generated here - https://github.com/JuliaDiffEq/DiffEqBase.jl/pull/250/files#diff-78606e8d4e26065996ab8302a8d36405R150 and https://github.com/JuliaDiffEq/DiffEqBase.jl/pull/250/files#diff-78606e8d4e26065996ab8302a8d36405R156

**Comment by kanav99** [comment]
What I am confused about right now is whether this splitting would help us or not in the Parallel methods, though `nlsolver` generator functions I made here would definately help.

File: src/nlsolve/utils.jl (hunk 1 above)
Comment (ChrisRackauckas) [reply src/nlsolve/utils.jl]: I mean `iip_generate_W` should probably return J,W so it can be in the cache for the resizing algorithms.

Comment (kanav99) [reply src/nlsolve/utils.jl]: `iip_generate_W` is called inside the `iipnlsolve` function (the function which generates `nlsolver`). So even if J is returned, J is not used in generating nlsolver. We can do this instead - make `iip_generate_W` return `J,W`, call `iip_generate_W` in `alg_cache` and pass the generated `W` in `iipnlsolve` function. That would be much safer I guess.

**Comment by kanav99** [comment]
Something weird has happen with some of the dependencies, I get test failures (mostly where the function `all` is used) at random times.

**Comment by kanav99** [comment]
I think the PR is complete, just waiting for the upstream errors to be fixed. Once they are fixed, I will rerun the tests to make sure. Do not merge this, breaking changes.

File: src/nlsolve/utils.jl (hunk 1 above)
Comment (ChrisRackauckas) [reply src/nlsolve/utils.jl]: >We can do this instead - make iip_generate_W return J,W, call iip_generate_W in alg_cache and pass the generated W in iipnlsolve function.

Indeed that's what I am asking for.

**Comment by kanav99** [comment]
I have restored the original macros, so that this PR is not broken. I have also confirmed that you can have both a macro and function by same name, so thats not a problem too.

File: src/nlsolve/utils.jl [hunk 2]
```diff
@@ -233,3 +233,110 @@ DiffEqBase.@def oopnlsolve begin
...
+
+DiffEqBase.@def getiipnlsolvefields begin
+  @unpack z,dz,tmp,k = nlsolver
+  b = nlsolver.ztmp
+  fsalfirst = zero(rate_prototype)
+
+  if alg.nlsolve isa NLNewton
+    nf = nlsolve_f(f, alg)
+    islin = f isa Union{ODEFunction,SplitFunction} && islinear(nf.f)
+    if islin
```
Comment (ChrisRackauckas) [inline src/nlsolve/utils.jl]: All of these caches will need to be nlsolver local in order to be thread-safe.

Comment (kanav99) [reply src/nlsolve/utils.jl]: Realized now that if I do this, the code will break. Or should I make the new entries in nlsolver default to nothing? I am not sure of the side effects

**Comment by kanav99** [comment]
All the remaining fields have been moved to nlsolver. Interface tests pass for the new branch, but I will inform you once I confirm that master of OrdinaryDiffEq doesn't break too

**Pull request merged**